internal/pkg/fleetapi/acker/retrier/retrier.go

```diff
@@ -0,0 +1,213 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package retrier
+
+import (
+	"context"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/elastic/elastic-agent/internal/pkg/core/backoff"
+	"github.com/elastic/elastic-agent/internal/pkg/core/logger"
+	"github.com/elastic/elastic-agent/internal/pkg/fleetapi"
+)
+
+const (
+	defaultMaxRetries = 5
+
+	defaultInitialRetryInterval = 1 * time.Minute
+	defaultMaxRetryInterval     = 5 * time.Minute
+)
+
+// BatchAcker provider interface, implemented by fleet acker
+type BatchAcker interface {
+	AckBatch(ctx context.Context, actions []fleetapi.Action) (*fleetapi.AckResponse, error)
+}
+
+// Option Retrier option function
+type Option func(*Retrier)
+
+// Retrier implements retrier for actions acks
+type Retrier struct {
+	log   *logger.Logger
+	acker BatchAcker // AckBatch provider
+
+	initialRetryInterval time.Duration // initial retry interval
+	maxRetryInterval     time.Duration // max retry interval
+	maxRetries           int           // configurable maxNumber of retries per action
+
+	actions []fleetapi.Action // pending actions
+	mx      sync.Mutex
+
+	kickCh chan struct{} // signal channel to kickoff retry loop if not running
+
+	doneCh chan struct{} // signal channel when retry loop is done
+}
+
+// New creates new instance of retrier
+func New(acker BatchAcker, log *logger.Logger, opts ...Option) *Retrier {
+	r := &Retrier{
+		acker:                acker,
+		log:                  log,
+		initialRetryInterval: defaultInitialRetryInterval,
+		maxRetryInterval:     defaultMaxRetryInterval,
+		maxRetries:           defaultMaxRetries,
+		kickCh:               make(chan struct{}, 1),
+		doneCh:               make(chan struct{}, 1),
+	}
+	for _, opt := range opts {
+		opt(r)
+	}
+	return r
+}
+
+// WithInitialRetryInterval configures retrier with initial retry delay provided
+func WithInitialRetryInterval(dur time.Duration) Option {
+	return func(f *Retrier) {
+		f.initialRetryInterval = dur
+	}
+}
+
+// WithMaxRetryInterval configures retrier with max retry interval provided
+func WithMaxRetryInterval(dur time.Duration) Option {
+	return func(f *Retrier) {
+		f.maxRetryInterval = dur
+	}
+}
+
+// WithMaxAckRetries configures max retries provided
+// The number of retries left is tracked per action
+func WithMaxAckRetries(maxRetries int) Option {
+	return func(f *Retrier) {
+		f.maxRetries = maxRetries
+	}
+}
+
+// Done signals when retry loop is done, useful for testing
+func (r *Retrier) Done() <-chan struct{} {
+	return r.doneCh
+}
+
+// Run runs retrier loop
+func (r *Retrier) Run(ctx context.Context) {
+	for {
+		select {
+		case <-r.kickCh:
+			r.runRetries(ctx)
+		case <-ctx.Done():
+			r.log.Debug("ack retrier: exit on context cancelled")
```

Comment:
Might be better to add the actual error reason, possible context closed due to a timeout.

```suggestion
			r.log.Debug("ack retrier: exit on %w", ctx.Error())
```